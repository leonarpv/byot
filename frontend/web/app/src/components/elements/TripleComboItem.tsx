import * as React from 'react';
import {
  BoxProps,
  ListItem as ListItemBase,
  ListItemProps,
  ListItemText,
  makeStyles,
  Theme,
} from '@material-ui/core';
import {WithStyles} from '@byot-frontend/web-common/src/types/WithStyles';
import {Patch} from '@byot-frontend/web-common/src/components/elementary/patch/Patch';
import {CSSProperties} from '@material-ui/styles';

type Props<P> = WithStyles<typeof styles> &
  ListItemProps & {
    title?: React.ReactNode;
    image?: string;
    imagePlaceholder?: boolean | React.ReactNode;
    description?: React.ReactNode;
    component?: React.ComponentType<P>;
    singleLine?: boolean;
    transparent?: boolean;
  } & P;

const styles = (theme: Theme) => ({
  root: {
    padding: theme.spacing(2),
  },
  image: (props: Props<any>) => ({
    borderRadius: theme.shape.borderRadius,
    width: theme.spacing(8),
    height: theme.spacing(8),
    minWidth: theme.spacing(8),
    overflow: 'hidden',
    backgroundSize: 'cover',
    backgroundPosition: 'center',
    marginRight: theme.spacing(2),
    ...(props.image && {backgroundImage: `url(${props.image})`}),
  }),
  primaryText: (props: Props<any>) => ({
    fontWeight: 700,
    ...(props.singleLine &&
      ({
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
      } as CSSProperties)),
  }),
  textRoot: (props: Props<any>) => ({
    margin: props.singleLine && 0,
  }),
});
const useStyles = makeStyles(styles);

export const TripleComboItem = <P extends object = object>(props: Props<P>) => {
  const styles = useStyles(props);
  const {title, description, imagePlaceholder, image, transparent, ...baseProps} = props;
  return (
    <ListItemBase
      alignItems="flex-start"
      {...(baseProps as any)}
      component={props.transparent ? undefined : p => <Patch component={props.component} {...p} />}
      classes={{root: styles.root, ...baseProps.classes}}>
      {(props.image || props.imagePlaceholder) && (
        <div className={styles.image}>
          {typeof props.imagePlaceholder !== 'boolean' && !props.image && props.imagePlaceholder}
        </div>
      )}
      {(props.title || props.description) && (
        <ListItemText
          classes={{primary: styles.primaryText, root: styles.textRoot}}
          primary={props.title}
          secondary={props.description}
        />
      )}
      {props.children}
    </ListItemBase>
  );
};
