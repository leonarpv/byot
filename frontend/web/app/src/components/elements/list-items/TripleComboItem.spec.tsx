import * as React from 'react';
import {render} from '@testing-library/react';
import {TripleComboItem} from './TripleComboItem';

describe('<TripleComboItem/>', () => {
  it('should render with different props', () => {
    expect(
      render(<TripleComboItem title="Something" description="description" />).container
    ).toMatchSnapshot();
    expect(
      render(<TripleComboItem title="Something" description="description" transparent />).container
    ).toMatchSnapshot();
    expect(render(<TripleComboItem title="Something" singleLine />).container).toMatchSnapshot();
    expect(
      render(<TripleComboItem title="Something" description="description" image="image" />).container
    ).toMatchSnapshot();
    expect(
      render(<TripleComboItem title="Something" description="description" imagePlaceholder />).container
    ).toMatchSnapshot();
  });
});
