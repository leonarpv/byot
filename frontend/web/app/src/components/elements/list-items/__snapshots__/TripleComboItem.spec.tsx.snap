// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripleComboItem/> should render with different props 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-9 MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-alignItemsFlexStart makeStyles-root-8"
  >
    <div
      class="MuiListItemText-root makeStyles-textRoot-4 makeStyles-textRoot-7 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-3 makeStyles-primaryText-6 MuiTypography-body1 MuiTypography-displayBlock"
      >
        Something
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        description
      </p>
    </div>
  </div>
</div>
`;

exports[`<TripleComboItem/> should render with different props 2`] = `
<div>
  <li
    class="MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-alignItemsFlexStart"
  >
    <div
      class="MuiListItemText-root makeStyles-textRoot-4 makeStyles-textRoot-12 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-3 makeStyles-primaryText-11 MuiTypography-body1 MuiTypography-displayBlock"
      >
        Something
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        description
      </p>
    </div>
  </li>
</div>
`;

exports[`<TripleComboItem/> should render with different props 3`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-16 MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-alignItemsFlexStart makeStyles-root-8"
  >
    <div
      class="MuiListItemText-root makeStyles-textRoot-4 makeStyles-textRoot-15"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-3 makeStyles-primaryText-14 MuiTypography-body1 MuiTypography-displayBlock"
      >
        Something
      </span>
    </div>
  </div>
</div>
`;

exports[`<TripleComboItem/> should render with different props 4`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-20 MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-alignItemsFlexStart makeStyles-root-8"
  >
    <div
      class="makeStyles-image-2 makeStyles-image-17"
    />
    <div
      class="MuiListItemText-root makeStyles-textRoot-4 makeStyles-textRoot-19 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-3 makeStyles-primaryText-18 MuiTypography-body1 MuiTypography-displayBlock"
      >
        Something
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        description
      </p>
    </div>
  </div>
</div>
`;

exports[`<TripleComboItem/> should render with different props 5`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-24 MuiListItem-root makeStyles-root-1 MuiListItem-gutters MuiListItem-alignItemsFlexStart makeStyles-root-8"
  >
    <div
      class="makeStyles-image-2 makeStyles-image-21"
    />
    <div
      class="MuiListItemText-root makeStyles-textRoot-4 makeStyles-textRoot-23 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary makeStyles-primaryText-3 makeStyles-primaryText-22 MuiTypography-body1 MuiTypography-displayBlock"
      >
        Something
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        description
      </p>
    </div>
  </div>
</div>
`;
